File: 03.CPP/0209.cpp
#include <iostream>
#include <vector>
#include <limits>

class Solution {
public:
    int minSubArrayLen(int target, std::vector<int>& nums) {
        int n = nums.size();

        int left = 0;
        int windowSum = 0;
        int minLen = std::numeric_limits<int>::max();

        for (int right = 0; right < n; right++) {
            windowSum += nums[right];

            while (windowSum >= target) {
                int currLen = right - left + 1;

                if (currLen < minLen) {
                    minLen = currLen;
                }

                windowSum -= nums[left];
                left++;
            }
        }

        if (minLen == std::numeric_limits<int>::max()) {
            return 0;
        } else {
            return minLen;
        }
    }
};

int main() {
    Solution sol;

    int target = 7;
    std::vector<int> nums = {2, 3, 1, 2, 4, 3};

    std::cout << sol.minSubArrayLen(target, nums) << "\n";
    return 0;
}

File: 01.Python/0209.py
from typing import List

class Solution:
    def minSubarrayLen(self, target: int, nums: List[int]) -> int:
        n = len(nums)

        left = 0
        windowSum = 0
        minLen = float('inf')

        for right in range(n):
            windowSum += nums[right]

            while windowSum >= target:
                currLen = right - left + 1
                if currLen < minLen:
                    minLen = currLen

                windowSum -= nums[left]
                left += 1

        if minLen == float('inf'):
            return 0
        else:
            return minLen


sol = Solution()
target = 7
nums = [2, 3, 1, 2, 4, 3]

print(sol.minSubarrayLen(target, nums))